### DISCUSSION TITLE
How to do custom build.build and build_ext.build_ext classes in Python 3.12 (i.e., without the stdlib distutils)?

### DISCUSSION THREAD

**Discussion by mcepl** [pypa/setuptools#4062] [Q&A] [answered]
I have this rather gigantic [`setup.py` for M2Crypto](https://gitlab.com/m2crypto/m2crypto/-/blob/master/setup.py) with this custom classes for building Python extensions:

```python
from distutils.command import build
from distutils.command.clean import clean
from distutils.dir_util import mkpath

import setuptools
from setuptools.command import build_ext

# ... many lines skipped over

class _M2CryptoBuild(build.build):
    """Enable swig_opts to inherit any include_dirs settings made elsewhere."""

    user_options = build.build.user_options + \
        [('openssl=', 'o', 'Prefix for openssl installation location')] + \
        [('bundledlls', 'b', 'Bundle DLLs (win32 only)')]

    def initialize_options(self):
        """Overload to enable custom openssl settings to be picked up."""
        build.build.initialize_options(self)
        self.openssl = None
        self.bundledlls = None


class _M2CryptoBuildExt(build_ext.build_ext):
    """Enable swig_opts to inherit any include_dirs settings made elsewhere."""

    user_options = build_ext.build_ext.user_options + \
        [('openssl=', 'o', 'Prefix for openssl installation location')] + \
        [('bundledlls', 'b', 'Bundle DLLs (win32 only)')]

    def initialize_options(self):
        """Overload to enable custom openssl settings to be picked up."""
        build_ext.build_ext.initialize_options(self)
        self.openssl = None
        self.bundledlls = None

    def finalize_options(self):
        # type: (None) -> None
        """Append custom openssl include file and library linking options."""
        build_ext.build_ext.finalize_options(self)
        self.openssl_default = None

        # ... a lot of very complicated configuration to make build work
        # on multiple platforms and with different architectures ...

    def run(self):
        """
        On Win32 platforms include the openssl dll's in the binary packages
        """

        # ... more complicated stuff to make building on Win32 possible ...

        build_ext.build_ext.run(self)


class Clean(clean):
    def __init__(self, dist):
        clean.__init__(self, dist)

    def initialize_options(self):
        clean.initialize_options(self)
        self.all = True

    def finalize_options(self):
        clean.finalize_options(self)

    def run(self):
        clean.run(self)
        garbage_list = [
            os.path.join('src', 'M2Crypto', '*.so'),
            os.path.join('src', 'M2Crypto', '*.pyd'),
            os.path.join('src', 'M2Crypto', '*.dll')
        ]
        for p in garbage_list:
            for f in glob.glob(p):
                if os.path.exists(f):
                    os.unlink(f)
```

Now when I just try to import the same stuff from new locations:

```python
if sys.version_info[:2] < (3, 10):
    from distutils.command import build
    from distutils.command.clean import clean
    from distutils.dir_util import mkpath
else:
    from setuptools.command import build
    from setuptools._distutils.command.clean import clean

from setuptools.command import build_ext
```

I get dreaded `AttributeError: 'MinimalDistribution' object has no attribute 'set_defaults'` exception:

```
tumbleweed-pkg~/r/m2crypto (setuptools_py312 *)$ python3 setup.py clean 
Traceback (most recent call last):
  File "/home/matej/archiv/knihovna/repos/m2crypto/setup.py", line 347, in <module>
    setuptools.setup(
  File "/usr/lib/python3.11/site-packages/setuptools/__init__.py", line 102, in setup
    _install_setup_requires(attrs)
  File "/usr/lib/python3.11/site-packages/setuptools/__init__.py", line 70, in _install_setup_requires
    dist = MinimalDistribution(attrs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/setuptools/__init__.py", line 54, in __init__
    self.set_defaults._disable()
    ^^^^^^^^^^^^^^^^^
AttributeError: 'MinimalDistribution' object has no attribute 'set_defaults'
tumbleweed-pkg~/r/m2crypto (setuptools_py312 *)$ python3 setup.py clean
Traceback (most recent call last):
  File "/home/matej/archiv/knihovna/repos/m2crypto/setup.py", line 349, in <module>
    setuptools.setup(
  File "/usr/lib/python3.11/site-packages/setuptools/__init__.py", line 102, in setup
    _install_setup_requires(attrs)
  File "/usr/lib/python3.11/site-packages/setuptools/__init__.py", line 70, in _install_setup_requires
    dist = MinimalDistribution(attrs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/setuptools/__init__.py", line 54, in __init__
    self.set_defaults._disable()
    ^^^^^^^^^^^^^^^^^
AttributeError: 'MinimalDistribution' object has no attribute 'set_defaults'
tumbleweed-pkg~/r/m2crypto (setuptools_py312 *)$ 
```

I was looking at https://github.com/pypa/setuptools/issues/3743 and the solution there (“just import directly from the stdlib distutils”) is obviously not applicable any more for 3.12+. In the same time I cannot find anywhere any official (or non-official) documentation how to create custom classes with setuptools.

Could anybody point me in the right direction please?

**Comment by abravalheri** [comment, marked as answer]
Hi @mcepl,

My understanding of how this works is the following:

1. `setuptools._distutils` is private for a reason and should not be imported.
   Indeed as discussed in #3743, importing directly from `setuptools._distutils` will actively break your code.
   This happens because `setuptools._distutils` is not implemented in such a way that it can be imported directly, but instead it needs to go through a custom [`MetaPathFinder`](https://docs.python.org/3/reference/import.html#finders-and-loaders) implemented by setuptools.
3. The statement in:
   > ... the solution there (“just import directly from the stdlib distutils”) is obviously not applicable any more for 3.12+.

   is not "completely" valid. Note that the solution presented in the other discussion was never "import directly from the ***stdlib*** distutils", but rather "import directly from distutils". Importing directly from `distutils` is perfectly possible as long as you have setuptools installed.
   It is expected to go away someday, however, once setuptools subsumes it completely. When this is going to happen is undetermined, but when it is announced I am sure setuptools will inform the users accordingly.

Another note is that calling `python setup.py clean` directly is no longer supported by setuptools (running any command with [`python setup.py` is deprecated](https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html)), so it might make less sense to implement it nowadays.

There is documentation on how to implement custom build sub-commands in https://setuptools.pypa.io/en/latest/userguide/extension.html and an example in https://github.com/pypa/setuptools/discussions/3762, but arbitrary commands to run in the CLI are [no longer supported by setuptools](https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html).

If you need custom runnable commands, maybe you can try tools like [tox](https://tox.wiki/en/4.11.3/), [nox](https://nox.thea.codes/en/stable/), [invoke](https://www.pyinvoke.org), [doit](https://pydoit.org/contents.html), etc...

**Reply by mcepl** [reply]
> If you need custom runnable commands, maybe you can try tools like tox, nox, invoke, doit, etc...

Yeah, because why should developers be able to actually develop just with standard library, when we can force them to use twenty libraries, which are constantly out-of-date from each other and certainly not supported for long run (mind you, M2Crypto still works on RHEL-7 with Python 2.7)?

Thank you for the response, it is certainly factually correct, and just hate Python world a little bit more.
